### PR TITLE
Remove pkey_downgrade from PKCS7 code

### LINKED ISSUE #12991: Downgrade of key in PKCS7_SIGNER_INFO_set()
We are deliberately downgrading a key in PKCS7_SIGNER_INFO_set() which suggests missing functionality. Do we need to do something?

Reference:

https://github.com/openssl/openssl/blob/a48309cb5c58be8fa310608124925d4729664cf8/crypto/pkcs7/pk7_lib.c#L316-L320

### PR DESCRIPTION
**Pull request by slontis**
Fixes #12991

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
I assume I am missing something here based on what @levitte has commented on in issue #12991
I tried to do something similar to the CMS case.

**Comment by slontis** [comment]
Travis errors are not related to this PR.

**Comment by slontis** [comment]
ping

File: crypto/pkcs7/pk7_ecdsa.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/pkcs7.h>
+#include <openssl/x509.h>
+#include "crypto/asn1.h"
+#include "pk7_local.h"
+
+int pkcs7_ecdsa_dsa_sign(PKCS7_SIGNER_INFO *si, int verify)
```
Comment (t-j-h) [inline crypto/pkcs7/pk7_ecdsa.c:15]: Perhaps pkcs7_ecdsa_dsa_sign should be pkcs7_ecdsa_or_dsa_sign to be clearer?
And actually given it is both sign and verify calling it just "sign" seems wrong.
Perhaps pkcs7_ecdsa_or_dsa_sign_verify?

And looking at all the code we are using the three new functions in only one place - so why are we adding these additional global symbols for a single user and each of these functions is rather short/simple?

Comment (slontis) [reply crypto/pkcs7/pk7_ecdsa.c:15]: you could also argue that sign_verify is not a good name either.. They look more like a resolve stage for doing those particular operations..

Comment (t-j-h) [reply crypto/pkcs7/pk7_ecdsa.c:15]: Indeed. "resolve" or "x" or something less misdirecting would be better IMHO - but if only used once then they don't need to be separate functions - or name them for where they are called from ... or "bridge" in the name or ...

Comment (slontis) [reply crypto/pkcs7/pk7_ecdsa.c:15]: will collapse into pkcs7_lib.c - It is not a large source file (in comparison to others).

**Comment by levitte** [comment]
Except for the algorithms that we know we have implementations for, this drops all support for provided key types.
Since this is all about AlgorithmIdentifier, I'll just mention that there is an OSSL_PARAM key to ask the providers for it, "algorithm-id". which is currently only implemented for signatures (`OSSL_SIGNATURE_PARAM_ALGORITHM_ID`), but the same mechanism can be used anywhere we need an AlgorithmIdentifier.  The response on this parameter is expected to be a DER encoded octet string.

Furthermore, I'd suggest a study of this:

https://github.com/gost-engine/engine/blob/711b83cba2fb9d5615e83ce17ccd1fbd66d4b9db/gost_ameth.c#L438-L445
https://github.com/gost-engine/engine/blob/711b83cba2fb9d5615e83ce17ccd1fbd66d4b9db/gost_ameth.c#L456-L466

**Comment by slontis** [comment]
> furthermore, I'd suggest a study of this:

@levitte I am still missing something...

i.e it tries the EVP_PKEY_is_a() key types followed by
ret = pkey->ameth->pkey_ctrl(pkey, ASN1_PKEY_CTRL_PKCS7_ENCRYPT, 0, p7i);

This looks like the same thing as is done in the CMS code that was converted as far as I can see..
So are you saying the CMS case is also not working?
e.g see cms_sd_asn1_ctrl() in cms_sd.c??

**Comment by levitte** [comment]
I just looked at the code in `asn1_ctrl` in `cms_sd.c`, and yeah, there's a problem there too...  `EVP_PKEY_is_a()` can answer correctly on legacy EVP_PKEYs as well, so this means that functions like `cms_rsa_sign()` will be called regardless of if the EVP_PKEY internal key is legacy or provided...

We have internal macros, `evp_pkey_is_legacy()` and `evp_pkey_is_provided()`, that should be used to decide what code path to use.

**Comment by mattcaswell** [comment]
> I just looked at the code in asn1_ctrl in cms_sd.c, and yeah, there's a problem there too... EVP_PKEY_is_a() can answer correctly on legacy EVP_PKEYs as well, so this means that functions like cms_rsa_sign() will be called regardless of if the EVP_PKEY internal key is legacy or provided...

So what? I don't see anything in cms_rsa_sign() that is legacy or provider specific.

**Comment by mattcaswell** [comment]
The legacy fallback code in `cms_sd_asn1_ctrl` is only intended for cases where we have a key type that we don't know anything about. If we *know* we have RSA then we should call `cms_rsa_sign()` regardless.

**Comment by levitte** [comment]
> So what? I don't see anything in cms_rsa_sign() that is legacy or provider specific.

... and in comes an ENGINE implementing RSA.

**KABOOM**

**Comment by mattcaswell** [comment]
> KABOOM

Err. No? Why?

**Comment by mattcaswell** [comment]
`cms_rsa_sign()` just makes standard EVP calls and assume nothing about the internal implementation. I see nothing in it that would cause an ENGINE implementation to fail.

**Comment by levitte** [comment]
> Err. No? Why?

You're right, I wasn't clear enough.  What if that ENGINE supports the controls that are now not called at all?  What if those controls are somehow meaningful in ways we simply cannot predict?

**Comment by mattcaswell** [comment]
> You're right, I wasn't clear enough. What if that ENGINE supports the controls that are now not called at all? What if those controls are somehow meaningful in ways we simply cannot predict?

So you're saying that these ctrls might have weird side-effects nothing to do with the ctrl itself? The intended semantics of these ctrls is essentially stateless. If ENGINEs are hanging other stuff off of these then I think that's an abuse of the ctrl. I don't think we should be concerned about that.

**Comment by levitte** [comment]
I'll review this PR again in light of what's been discussed, 'cause there were aspects I hadn't seen before.

> If ENGINEs are hanging other stuff off of these then I think that's an abuse of the ctrl. I don't think we should be concerned about that.

Considering the utter lack of documentation on the controls called by PKCS#7 and CMS code, it's very hard to say what's abuse and what isn't.  Basically, libcrypto is handing over pointers to certain structures and ENGINEs are left doing something "sensible" with them, and "sensible" is pretty much undefined.

**Comment by mattcaswell** [comment]
> Considering the utter lack of documentation on the controls called by PKCS#7 and CMS code, it's very hard to say what's abuse and what isn't. Basically, libcrypto is handing over pointers to certain structures and ENGINEs are left doing something "sensible" with them, and "sensible" is pretty much undefined.

True, but I think this interpretation is perfectly reasonable. I think the chance of actually breaking something is very small. GOST aside, I'm willing to bet very few engines actually implement this ctrl  - and if they do they are unlikely to be broken by this change. I see the risk as sufficiently small that this approach is perfectly ok for a major release.

**Comment by levitte** [comment]
However, I still think that the CMS code should be used as a point of study.  After all, *our* RSA etc implementations only touch one of the algos in the `ASN1_PKEY_CTRL_PKCS7_SIGN` ctrl, while the GOST engine touches both...  we might want to think once more about how that translates into provider terms, should the hash algorithmID come from somewhere else or is that something that's pre-determined elsewhere?

**Comment by levitte** [comment]
> I'm willing to bet very few engines actually implement this ctrl

I hope you're right.  To be safe, I had a look at engine-pkcs11, and it seems that it indeed doesn't implement those (or any EVP_PKEY_ASN1_METHOD), so we're safe there.

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**